pkg/controller/metrics/provision_underway_collector_test.go

```diff
@@ -566,6 +570,279 @@ func TestProvisioningUnderwayInstallRestartsCollector(t *testing.T) {
+			require.NoError(t, c.List(context.TODO(), cdList))
+			for _, cd := range cdList.Items {
+				cd.ObjectMeta.Finalizers = nil
+				require.NoError(t, c.Update(context.TODO(), &cd))
+			}
+
+			ch2 := make(chan prometheus.Metric)
+			go func() {
+				collect.Collect(ch2)
+				close(ch2)
+			}()
+			for sample := range ch2 {
+				var d dto.Metric
+				require.NoError(t, sample.Write(&d))
+				got = append(got, metricPretty(d))
+			}
+			assert.Equal(t, test.expected, got)
+		})
+	}
+
+}
+
+func TestClusterSyncFailingCollector(t *testing.T) {
+	scheme := runtime.NewScheme()
+	hiveintv1alpha1.AddToScheme(scheme)
+
+	cases := []struct {
+		name string
+
+		existing []runtime.Object
+		min      time.Duration
+
+		expected []string
+	}{
+		{
+			name: "clustersync did not pass threshold",
+			existing: []runtime.Object{
+				testcs.FullBuilder("test-namespace", "test-name", scheme).Options(FailingSince(time.Now())).Build(),
+			},
+			min:      1 * time.Hour,
+			expected: []string(nil),
+		},
+		{
+			name: "clustersync passed threshold",
+			existing: []runtime.Object{
+				testcs.FullBuilder("test-namespace", "test-name", scheme).Options(FailingSince(time.Now())).Build(),
+			},
+			min:      0 * time.Hour,
+			expected: []string{"namespaced_name = test-namespace/test-name"},
+		},
+		{
+			name:     "no clustersync",
+			existing: nil,
+			min:      1 * time.Hour,
+			expected: []string(nil),
+		},
+	}
+	for _, test := range cases {
+		t.Run("test", func(t *testing.T) {
+			c := fake.NewClientBuilder().WithScheme(scheme).WithRuntimeObjects(test.existing...).Build()
+
+			collect := newClusterSyncFailingCollector(c, test.min)
+
+			descCh := make(chan *prometheus.Desc)
+			go func() {
+				for range descCh {
+				}
+			}()
+			ch := make(chan prometheus.Metric)
+			go func() {
+				collect.Collect(ch)
+				close(ch)
+			}()
+
+			var got []string
+			for sample := range ch {
+				var d dto.Metric
+				require.NoError(t, sample.Write(&d))
+				got = append(got, metricPretty(d))
+			}
+			assert.Equal(t, test.expected, got)
+
+		})
+	}
+}
+
+func TestDeletedClusterSyncFailingCollector(t *testing.T) {
+	scheme := runtime.NewScheme()
+	hiveintv1alpha1.AddToScheme(scheme)
+
+	cases := []struct {
+		name string
+
+		existing []runtime.Object
+		min      time.Duration
+
+		expected []string
+	}{
+		{
+			name: "clustersync deleted",
+			existing: []runtime.Object{
+				testcs.FullBuilder("test-namespace", "test-name", scheme).Options(FailingSince(time.Now())).Build(),
+			},
+			min:      0 * time.Hour,
+			expected: []string(nil),
+		},
+	}
+	for _, test := range cases {
+		t.Run("test", func(t *testing.T) {
+			c := fake.NewClientBuilder().WithScheme(scheme).WithRuntimeObjects(test.existing...).Build()
+
+			collect := newClusterSyncFailingCollector(c, test.min)
+
+			descCh := make(chan *prometheus.Desc)
+			go func() {
+				for range descCh {
+				}
+			}()
+
+			csList := &hiveintv1alpha1.ClusterSyncList{}
```

Comment:
Yeah I think there are channels opened but not read from and not closed? Or maybe I just don't understand it well enough